### ISSUE TITLE
More DCE wrt case object unapply

### ISSUE THREAD

**Issue by japgolly** [scala-js/scala-js#2396]
I was pleasantly surprised to see that entire `case objects` are eliminated when unused.

DCE doesn't reduce this snippet however:

```scala
sealed abstract class Blah
case object ABC extends Blah
case object XYZ extends Blah

def blahName(b: Blah): String =
  b match {
    case ABC => "abc"
    case XYZ => "xyz"
  }

def main(): Unit =
  println(blahName(ABC))
```

Because we know the following:
* With the exception of `XYZ.unapply` in a `case` clause, `XYZ` is unused.
* `XYZ.unapply` isn't overridden or overloaded.

I believe DCE could eliminate:
* The `XYZ` obejct itself.
* All `case` clauses referencing `XYZ`.

There by reducing the above snippet to:
```scala
sealed abstract class Blah
case object ABC extends Blah
// case object XYZ extends Blah

def blahName(b: Blah): String =
  b match {
    case ABC => "abc"
    // case XYZ => "xyz"
  }

def main(): Unit =
  println(blahName(ABC))
```

**Comment by japgolly** [comment]
Another instance of use that could be eliminated is `_: XYZ.type` in a `case` clause.

Will update head comment.

**Comment by gzm0** [comment]
What happens if you mark `blahName` as `@inline`?

**Comment by gzm0** [comment]
@sjrd I guess we could teach the optimizer to check for existence of instances when doing `isInstanceOf` checks, which should fold the if branch. The question is how much impact this will have on invalidation.

**Comment by sjrd** [comment]
Well here the thing the optimizer sees is `if (XYZ == b)`. Constant-folding that to `false` is hard.

**Comment by japgolly** [comment]
> What happens if you mark blahName as @inline?

Just gave it a try with `@inline`; it doesn't eliminate `XYZ`.

> Well here the thing the optimizer sees is `if (XYZ == b)`. Collapsing that to `false` is hard.

I don't know if this is a naive view but I imagine this optimisation could be implemented by something like:
* change usage analysis of case objects so that references (usages) are recorded in two distinct buckets: one for value inspection usage like `case` clauses, equality or reference comparison, instanceOf checking; and a bucket for everything else.
* DCE₁: remove objects without any references in usage-bucket-2.
* DCE₂: for all references in all removed objects' usage-bucket-1s, remove `case` clauses, collapse equality tests to constant `false`.
* Today's DCE runs and removes dead branches which now includes those introduced above.

Would that work? Is the reason you foresee this being hard due to logic (in that above is too naive) or in implementation?

**Comment by sjrd** [comment]
Not really. What we want to get rid of is `Obj == x` (Scala's equality) for pattern matching. But `==` is not always reference equality. For example `Nil == Vector.empty`.

After a round of optimization, `==` simplifies to `===` (aka `eq`) when that's the actual implementation. But after the optimizer we only have one round of dce left, which is not enough to implement the rest of your "plan" (steps 2 and following). In particular what you call "DCE 2" actually needs a round of optimization, not dce.

**Comment by gzm0** [comment]
Note to self: It seems like
- Ignoring `_ === LoadModule(_)` trees in reachability analysis and
- Folding the above to false if the module isn't instantiated

Will give us the optimization we'd want. But the second step comes too late (Emitter would be our earliest phase). We can optimize the expression, but we cannot eliminate code that is unused because of it, providing limited value.

(Yes, this is what sjrd has been saying all along)

**Comment by gzm0** [comment]
I've had another (unsuccessful) pass at this. Sharing a negative result.

IIUC, as an approximation, the IR trees we are interested in are:

```js
x.equals;Ljava.lang.Object;Z(y)
```

where `x` is a value of
- module class type,
- without an override of `equals`

My thought was to special case this in the Analyzer (in `callMethod`) to
extract the fact that we are checking module class identity.

This seems to be possible; however, it does not solve the problem:
we cannot trace back what `x` is in the analyzer phase, so we cannot
eliminate / flag the `LoadModule` call (which we assigned to `x`) as "readonly".

It seems like the only other option (short of adding another optimizer-like phase)
would be to investigate compiler support (compile to a linker friendlier tree).
But IIUC that would not work in terms of separate compilation guarantees we need to offer (a module class needs to be able to override `equals` without re-compilation of usage sites).
